File: example-reactnative/ios/NativeEcrBridge/DiscoveryManager.swift
import Darwin
import Foundation
import OSLog

class DiscoveryManager {
    private let logger = Logger(subsystem: "com.paynl.ecr", category: "DiscoveryManager")

    private let whoIsMessage = "PAY.POS-WHO.IS".data(using: .utf8)
    private let iAmMessage = "PAY.POS-I.AM:"
    private let udpPort: UInt16 = 8889

    private var socket: Int32 = -1
    private var receiveTask: Task<Void, Never>?
    private var localSubnet: (base: String, count: Int)?
    private var localIP: String? // Store our own IP to skip it

    var isDiscovering: Bool {
        socket != -1
    }

    func startDiscovery() {
        logger.info("Start PAYNL discovery protocol")

        // Discover local subnet and IP
        guard let subnet = discoverLocalSubnet() else {
            logger.error("Failed to discover local subnet")
            return
        }
        localSubnet = subnet
        let localIp = localIP
        logger.info("Local subnet: \(subnet.base).0/24 (self: \(localIp ?? "unknown"))")

        guard startListening() else {
            logger.error("Failed to start listening")
            return
        }

        guard broadcastDiscovery() else {
            logger.error("Failed to send discovery broadcast")
            return
        }
    }

    func stopDiscovery() {
        guard isDiscovering else { return }
        closeSocket()
    }

    /// Discover the local subnet and our IP address
    private func discoverLocalSubnet() -> (base: String, count: Int)? {
        var ifaddr: UnsafeMutablePointer<ifaddrs>?

        guard getifaddrs(&ifaddr) == 0, let firstAddr = ifaddr else {
            return nil
        }

        defer { freeifaddrs(ifaddr) }

        let preferredOrder = ["en0", "en1", "bridge0", "en2", "en3"]
        var foundInterfaces: [(name: String, ip: String)] = []

        var current = firstAddr
        while true {
            if let addr = current.pointee.ifa_addr,
               addr.pointee.sa_family == UInt8(AF_INET)
            {
                let name = String(cString: current.pointee.ifa_name)

                let isExcluded = name == "lo0" ||
                    name.hasPrefix("pdp_ip") ||
                    name.hasPrefix("utun") ||
                    name.hasPrefix("ap") ||
                    name.hasPrefix("ipsec")

                if !isExcluded {
                    var hostname = [CChar](repeating: 0, count: Int(NI_MAXHOST))
                    let result = getnameinfo(
                        addr,
                        socklen_t(addr.pointee.sa_len),
                        &hostname,
                        socklen_t(hostname.count),
                        nil,
                        0,
                        NI_NUMERICHOST
                    )

                    if result == 0 {
                        let ipString = String(cString: hostname)

                        if !ipString.hasPrefix("169.254.") {
                            foundInterfaces.append((name, ipString))
                        }
                    }
                }
            }

            guard let next = current.pointee.ifa_next else { break }
            current = next
        }

        // Find preferred interface and store our IP
        for preferred in preferredOrder {
            if let entry = foundInterfaces.first(where: { $0.name == preferred }) {
                localIP = entry.ip // Store our IP
                logger.info("Using interface: \(entry.name) (\(entry.ip))")
                return parseSubnet(from: entry.ip)
            }
        }

        if let first = foundInterfaces.first {
            localIP = first.ip
            return parseSubnet(from: first.ip)
        }

        return nil
    }

    /// Parse IP address to extract subnet (e.g., "192.168.100.19" -> ("192.168.100", 254))
    private func parseSubnet(from ip: String) -> (base: String, count: Int)? {
        let parts = ip.split(separator: ".").compactMap { Int($0) }
        guard parts.count == 4 else { return nil }

        let base = "\(parts[0]).\(parts[1]).\(parts[2])"
        return (base, 254)
    }

    /// Send discovery message to all IPs in the subnet (except ourselves)
    private func broadcastDiscovery() -> Bool {
        guard let subnet = localSubnet, let data = whoIsMessage else {
            return false
        }

        logger.info("Sending discovery to all hosts in \(subnet.base).0/24")

        let startIP = 1
        let endIP = subnet.count

        var successCount = 0
        var skippedCount = 0

        for host in startIP ... endIP {
            let targetIP = "\(subnet.base).\(host)"

            // Skip our own IP address
            if targetIP == localIP {
                skippedCount += 1
                continue
            }

            if sendUnicast(data: data, to: targetIP, port: udpPort) {
                successCount += 1
            }
        }

        logger.info("Sent discovery to \(successCount) hosts (skipped self: \(skippedCount))")
        return successCount > 0
    }

    /// Send message to specific IP (unicast)
    private func sendUnicast(data: Data, to ip: String, port: UInt16) -> Bool {
        guard socket >= 0 else {
            return false
        }

        var addr = sockaddr_in()
        addr.sin_family = sa_family_t(AF_INET)
        addr.sin_port = port.bigEndian

        guard inet_pton(AF_INET, ip, &addr.sin_addr) == 1 else {
            return false
        }

        let sendResult = data.withUnsafeBytes { dataPtr in
            withUnsafePointer(to: &addr) { addrPtr in
                addrPtr.withMemoryRebound(to: sockaddr.self, capacity: 1) { sockaddrPtr in
                    sendto(socket, dataPtr.baseAddress, data.count, 0, sockaddrPtr, socklen_t(MemoryLayout<sockaddr_in>.size))
                }
            }
        }

        return sendResult >= 0
    }

    private func createSocket() -> Int32? {
        let sock = Darwin.socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP)
        guard sock >= 0 else {
            logger.error("Error creating socket: \(String(cString: strerror(errno)))")
            return nil
        }

        var reuseAddr: Int32 = 1
        setsockopt(sock, SOL_SOCKET, SO_REUSEADDR, &reuseAddr, socklen_t(MemoryLayout<Int32>.size))

        var reusePort: Int32 = 1
        setsockopt(sock, SOL_SOCKET, SO_REUSEPORT, &reusePort, socklen_t(MemoryLayout<Int32>.size))

        var timeout = timeval(tv_sec: 1, tv_usec: 0)
        setsockopt(sock, SOL_SOCKET, SO_RCVTIMEO, &timeout, socklen_t(MemoryLayout<timeval>.size))

        return sock
    }

    private func startListening() -> Bool {
        guard !isDiscovering else {
            logger.info("Already listening")
            return true
        }

        guard let sock = createSocket() else {
            logger.error("Failed to create socket")
            return false
        }
        socket = sock

        guard bindSocket(socket) else {
            closeSocket()
            return false
        }

        receiveTask = Task {
            self.receiveLoop()
        }

        return true
    }

    private func bindSocket(_ sock: Int32) -> Bool {
        var addr = sockaddr_in()
        addr.sin_family = sa_family_t(AF_INET)
        addr.sin_port = udpPort.bigEndian
        addr.sin_addr.s_addr = INADDR_ANY

        let bindResult = withUnsafePointer(to: &addr) { addrPtr in
            addrPtr.withMemoryRebound(to: sockaddr.self, capacity: 1) { sockaddrPtr in
                bind(sock, sockaddrPtr, socklen_t(MemoryLayout<sockaddr_in>.size))
            }
        }

        if bindResult < 0 {
            logger.error("Error binding socket: \(String(cString: strerror(errno)))")
            return false
        }

        logger.info("Socket bound to port 8889")
        return true
    }

    private func closeSocket() {
        if socket >= 0 {
            close(socket)
            socket = -1
        }
    }

    private func receiveLoop() {
        let bufferSize = 65536
        var buffer = [UInt8](repeating: 0, count: bufferSize)

        logger.info("Listening for responses")

        let startTime = Date()
        let discoveryTimeout: TimeInterval = 5.0

        while isDiscovering, socket >= 0 {
            if Date().timeIntervalSince(startTime) > discoveryTimeout {
                logger.info("Discovery timeout reached")
                break
            }

            if Task.isCancelled { break }

            var senderAddr = sockaddr_in()
            var addrLen = socklen_t(MemoryLayout<sockaddr_in>.size)

            let bytesRead = withUnsafeMutablePointer(to: &senderAddr) { addrPtr in
                addrPtr.withMemoryRebound(to: sockaddr.self, capacity: 1) { sockaddrPtr in
                    recvfrom(socket, &buffer, bufferSize, 0, sockaddrPtr, &addrLen)
                }
            }

            if bytesRead < 0 {
                if errno == EAGAIN || errno == EWOULDBLOCK {
                    continue
                }
                if errno == EINTR {
                    continue
                }
                logger.error("Receive error: \(String(cString: strerror(errno)))")
                break
            }

            if bytesRead == 0 { continue }

            let senderIP = String(cString: inet_ntoa(senderAddr.sin_addr))

            // Ignore responses from ourselves
            if senderIP == localIP {
                continue
            }

            let data = Data(buffer[0 ..< bytesRead])

            guard let message = String(data: data, encoding: .utf8) else {
                logger.warning("Invalid message received from \(senderIP)")
                continue
            }

            guard message.starts(with: iAmMessage) else {
                logger.warning("Unknown message from \(senderIP): \(message)")
                continue
            }

            let splitted = message.split(separator: ":")
            guard splitted.count == 3 else {
                logger.warning("Invalid message format from \(senderIP): \(message)")
                continue
            }

            let dict: [String: Any] = [
                "sourceIp": senderIP,
                "terminalCode": String(splitted[1]),
                "terminalName": String(splitted[2]),
            ]

            logger.info("Found terminal: \(String(splitted[1])) at \(senderIP)")
            EcrEventEmitter.emit(onDiscovery: dict)
        }

        stopDiscovery()
    }
}

File: examplereactnative/android/app/src/main/java/com/examplereactnative/ecr/DiscoveryManager.java
package com.examplereactnative.ecr;

import android.os.Handler;
import android.os.Message;
import android.util.Log;

import com.examplereactnative.ecr.models.PosTerminal;

import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.Inet4Address;
import java.net.InetAddress;
import java.net.NetworkInterface;
import java.net.SocketException;
import java.net.InetSocketAddress;
import java.net.SocketAddress;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class DiscoveryManager {
    private static final String TAG = "DiscoveryManager";

    private static final String WHO_IS_MESSAGE = "PAY.POS-WHO.IS";
    private static final String I_AM_MESSAGE = "PAY.POS-I.AM:";
    private static final int UDP_PORT = 8889;
    private static final int DISCOVERY_TIMEOUT_MS = 5000;
    private static final int BUFFER_SIZE = 65536;

    private static final String BROADCAST_ADDRESS = "255.255.255.255";

    private final Handler onDiscoverHandler;

    private DatagramSocket socket;
    private Thread receiveThread;
    private String localIP;
    private volatile boolean isDiscovering = false;

    public DiscoveryManager(Handler onDiscoverHandler) {
        this.onDiscoverHandler = onDiscoverHandler;
    }

    public boolean isDiscovering() {
        return isDiscovering;
    }

    public void startDiscovery() {
        Log.i(TAG, "Start PAYNL discovery protocol");

        // Discover local IP
        localIP = discoverLocalIP();
        if (localIP == null) {
            Log.e(TAG, "Failed to discover local IP");
            return;
        }
        Log.i(TAG, "Local IP: " + localIP);

        if (!startListening()) {
            Log.e(TAG, "Failed to start listening");
            return;
        }

        if (!broadcastDiscovery()) {
            Log.e(TAG, "Failed to send discovery broadcast");
            return;
        }
    }

    public void stopDiscovery() {
        if (!isDiscovering) {
            return;
        }
        closeSocket();
    }

    private String discoverLocalIP() {
        try {
            List<NetworkInterface> interfaces = Collections.list(NetworkInterface.getNetworkInterfaces());

            String[] preferredOrder = {"en0", "en1", "bridge0", "en2", "en3", "eth0", "wlan0"};

            List<NetworkInterfaceEntry> foundInterfaces = new ArrayList<>();

            for (NetworkInterface networkInterface : interfaces) {
                if (!networkInterface.isUp() || networkInterface.isLoopback()) {
                    continue;
                }

                // Skip excluded interface names
                String name = networkInterface.getName();
                if (name.startsWith("lo") ||
                        name.startsWith("pdp_ip") ||
                        name.startsWith("utun") ||
                        name.startsWith("ap") ||
                        name.startsWith("ipsec") ||
                        name.startsWith("rndis")) {
                    continue;
                }

                List<InetAddress> addresses = Collections.list(networkInterface.getInetAddresses());

                for (InetAddress addr : addresses) {
                    if (addr instanceof Inet4Address) {
                        String ipString = addr.getHostAddress();

                        // Skip link-local addresses
                        if (ipString.startsWith("169.254.")) {
                            continue;
                        }

                        foundInterfaces.add(new NetworkInterfaceEntry(name, ipString));
                    }
                }
            }

            // Find preferred interface and store our IP
            for (String preferred : preferredOrder) {
                for (NetworkInterfaceEntry entry : foundInterfaces) {
                    if (entry.name.equals(preferred)) {
                        Log.i(TAG, "Using interface: " + entry.name + " (" + entry.ip + ")");
                        return entry.ip;
                    }
                }
            }

            // Use first found interface
            if (!foundInterfaces.isEmpty()) {
                NetworkInterfaceEntry first = foundInterfaces.get(0);
                return first.ip;
            }

        } catch (SocketException e) {
            Log.e(TAG, "Error getting network interfaces", e);
        }

        return null;
    }

    private static class NetworkInterfaceEntry {
        String name;
        String ip;

        NetworkInterfaceEntry(String name, String ip) {
            this.name = name;
            this.ip = ip;
        }
    }

    /**
     * Send discovery broadcast to 255.255.255.255
     */
    private boolean broadcastDiscovery() {
        Log.i(TAG, "Sending discovery broadcast to " + BROADCAST_ADDRESS + ":" + UDP_PORT);

        byte[] data = WHO_IS_MESSAGE.getBytes();

        try {
            InetAddress broadcastAddress = InetAddress.getByName(BROADCAST_ADDRESS);
            DatagramPacket packet = new DatagramPacket(data, data.length, broadcastAddress, UDP_PORT);
            socket.send(packet);
            Log.i(TAG, "Broadcast sent successfully");
            return true;
        } catch (IOException e) {
            Log.e(TAG, "Error sending broadcast: " + e.getMessage());
            return false;
        }
    }

    private boolean startListening() {
        if (isDiscovering) {
            Log.i(TAG, "Already listening");
            return true;
        }

        if (!createSocket()) {
            Log.e(TAG, "Failed to create socket");
            return false;
        }

        if (!bindSocket()) {
            closeSocket();
            return false;
        }

        isDiscovering = true;

        receiveThread = new Thread(this::receiveLoop);
        receiveThread.start();

        return true;
    }

    private boolean createSocket() {
        try {
            // Create unbound socket
            socket = new DatagramSocket(null);

            // Set SO_REUSEADDR
            socket.setReuseAddress(true);

            // Enable SO_BROADCAST for broadcast messages
            socket.setBroadcast(true);

            // Set receive timeout to detect cancellation
            socket.setSoTimeout(1000);

            return true;
        } catch (IOException e) {
            Log.e(TAG, "Error creating socket: " + e.getMessage());
            return false;
        }
    }

    private boolean bindSocket() {
        try {
            SocketAddress localAddr = new InetSocketAddress(UDP_PORT);
            socket.bind(localAddr);
            Log.i(TAG, "Socket bound to port " + UDP_PORT);
            return true;
        } catch (IOException e) {
            Log.e(TAG, "Error binding socket: " + e.getMessage());
            return false;
        }
    }

    private void closeSocket() {
        isDiscovering = false;

        if (socket != null) {
            socket.close();
            socket = null;
        }

        if (receiveThread != null) {
            receiveThread.interrupt();
            receiveThread = null;
        }
    }

    private void receiveLoop() {
        byte[] buffer = new byte[BUFFER_SIZE];

        Log.i(TAG, "Listening for responses");

        long startTime = System.currentTimeMillis();
        long timeout = DISCOVERY_TIMEOUT_MS;

        while (isDiscovering && socket != null && !socket.isClosed()) {
            if (System.currentTimeMillis() - startTime > timeout) {
                Log.i(TAG, "Discovery timeout reached");
                break;
            }

            try {
                DatagramPacket packet = new DatagramPacket(buffer, buffer.length);

                socket.receive(packet);

                if (packet.getLength() == 0) {
                    continue;
                }

                InetAddress senderAddress = packet.getAddress();
                String senderIP = senderAddress.getHostAddress();

                // Ignore responses from ourselves
                if (senderIP.equals(localIP)) {
                    continue;
                }

                String message = new String(buffer, 0, packet.getLength(), "UTF-8");

                if (message == null) {
                    Log.w(TAG, "Invalid message received from " + senderIP);
                    continue;
                }

                if (!message.startsWith(I_AM_MESSAGE)) {
                    Log.w(TAG, "Unknown message from " + senderIP + ": " + message);
                    continue;
                }

                // Parse: "PAY.POS-I.AM:<terminalCode>:<terminalName>"
                String payload = message.substring(I_AM_MESSAGE.length());
                String[] splitted = payload.split(":");

                if (splitted.length != 2) {
                    Log.w(TAG, "Invalid message format from " + senderIP + ": " + message);
                    continue;
                }

                String terminalCode = splitted[0];
                String terminalName = splitted[1];

                Log.i(TAG, "Found terminal: " + terminalCode + " at " + senderIP);

                // Create discovery result
                PosTerminal result = new PosTerminal();
                result.sourceIp = senderIP;
                result.terminalCode = terminalCode;
                result.terminalName = terminalName;

                // Emit event
                Message msg = new Message();
                msg.obj = result;

                this.onDiscoverHandler.sendMessage(msg);
            } catch (IOException e) {
                if (!isDiscovering) {
                    // Socket was closed intentionally
                    break;
                }

                // Timeout is expected, continue
                continue;
            }
        }

        stopDiscovery();
    }
}